### PR TITLE
[WIP] Allow label names in labeled/unlabeled actions

### LINKED ISSUE #259: Allow users to specify specific label names in labeled events
I want to perform an action when a specific label is added:
```json
"rules": {
   "issues.labeled.question": "Moya/moya-peril/@org/someTask.ts"
}
```

### PR DESCRIPTION
**Pull request by SD10**
Resolves #259 

This is pointed at `enhancement/multi-rule-trigger` for now because it builds on its commits.

This allows users to specify a label name in a labeled or unlabeled action:
```
rules: {
   "pull_request.labeled.question": "dangerfile.js"
}
```

Things to note:
- I've encoded the $ character with 4 underscores because it's not mongo friendly and may be used in label names.
- Dots in label names were already handled

This is WIP because while writing it I noticed if a label name contains a comma for multi inline rules (#259) it won't split properly 😄

### CODE REVIEW

**Comment by SD10** [comment]
Probably wouldn't be able to support commas in a label name with the current syntax because it acts as the delimiter for multiple triggers.

Label Name: `a,comma`
Input: `issue.labeled.a,comma, issue.*`
Tokenized: `issue.labeled.a`, `comma`, `issue.*`

This results in missing the label run

Label Name: `question, issue.*`
Input: `issue.labeled.question, issue.*`
Tokenized: `issue.labeled.question`, `issue.*`

This results in an extra `issue.*` run. 

We would have to make the syntax be something like this:
```
rules: {
   "issue.labeled(this,name,works), issue.*"
}
```
Then I could probably strip them out first with a regex.

Personally, I would be fine with the limitation of not supporting commas in label names but LMK because changing it later introduces breaking changes.

**Comment by SD10** [comment]
a17f0df supports commas in labels but we can revert it if need be

**Comment by orta** [comment]
I’m afraid this is too specific, I don’t really want to include edge cases in the action/event parser like this - is rather we find a way to make it generic and work with any JSON keypath

**Comment by orta** [comment]
For example look at something like [keypather](https://www.npmjs.com/package/keypather) and extend that with something like a DSL, e.g:

```js
rules: {
   "pull_request.labeled(label == questioned)": "dangerfile.js"
}
```

Which uses the third `label` to pull out of the body JSON, and then compare that to the string "questioned". This could work for anything then:

```js
rules: {
   "pull_request.opened(number == 1000)": "happy_one_k.js",
   "status(state == success)": "merge_on_green.js"
}
```

**Comment by SD10** [comment]
Sorry, I misunderstood what you meant when you originally said using label names was too specific.

The reason I didn’t do the keypath solution is because I thought you would think it’s overkill. It could be really powerful IMO

Let me explore this a bit more

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by SD10** [comment]
The keypaths don't play that well with the most recent PR I submitted (allowing multi inline triggers).

There would be the following limitations:
- The keypath payload could contain the `(` or `)` delimiters.
- If the payload does contain those characters, it needs to be its own line.

This is because I would be using regular expressions to split up the inline triggers and (as far as I know) JavaScript does not support recursive regular expressions. (Which is needed to create a pattern that balances brackets)

Of course, the way around those limitations is to tokenize the input myself by writing a custom lexer for the keys of the rules JSON. Not sure how involved that would be and feels too heavy-handed for your tastes

TL;DR. Going to close this and will work on it more this Sunday

**Comment by orta** [comment]
Sure thing, FWIW, I'd rather have ^ than multi inline triggers FWIW 👍

**Comment by SD10** [comment]
Totally, I agree this would dwarf the multi trigger feature. Good to know you’d be fine reverting it

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by SD10** [comment]
Hey @orta,
I just stumbled upon my first developer job, so I will have less time for open source until I feel stable.
I know there's no pressure for contributing to open source, just wanted to keep you informed on the progression of this.

**Comment by orta** [comment]
Congrats! I got some time now actually, so I'll take a stab at it
